lib/location/method_wifi.c

```diff
@@ -0,0 +1,307 @@
+			break;
+		default:
+			break;
+		}
+	}
+}
+
+/******************************************************************************/
+
+static void method_wifi_positioning_work_fn(struct k_work *work)
+{
+	struct loc_location location_result = { 0 };
+	struct k_work_args *work_data = CONTAINER_OF(work, struct k_work_args, work_item);
+	const struct loc_wifi_config wifi_config = work_data->wifi_config;
+	int64_t starting_uptime_ms = work_data->starting_uptime_ms;
+	int err;
+
+	loc_core_timer_start(wifi_config.timeout);
+
+	err = method_wifi_scanning_start();
+	if (err) {
+		LOG_WRN("Cannot start WiFi scanning, err %d", err);
+		goto return_error;
+	}
+
+	k_sem_take(&wifi_scanning_ready, K_FOREVER);
+	if (running) {
+		struct rest_wifi_pos_request request;
+		struct rest_wifi_pos_result result;
+
+		if (!loc_utils_is_default_pdn_active()) {
+			/* No worth to start trying to fetch with the REST api over cellular.
+			 * Thus, fail faster in this case and safe the trying "costs".
+			 */
+			LOG_WRN("Default PDN context is NOT active, cannot retrieve a location");
+			err = -1;
+			goto return_error;
+		}
+
+		if (latest_scan_result_count > 1) {
+			/* Fill scanning results: */
+			request.wifi_scanning_result_count = latest_scan_result_count;
+
+			/* Calculate remaining time as a REST timeout (in mseconds): */
+			request.timeout_ms = ((wifi_config.timeout * 1000) -
+					   (k_uptime_get() - starting_uptime_ms));
+			if (request.timeout_ms < REST_WIFI_MIN_TIMEOUT_MS) {
+				if (request.timeout_ms < 0) {
+					/* No remaining time at all */
+					LOG_WRN("No remaining time left for requesting a position");
+					goto return_error;
+				}
+				request.timeout_ms = REST_WIFI_MIN_TIMEOUT_MS;
+			}
+
+			for (int i = 0; i < latest_scan_result_count; i++) {
+				strcpy(request.scanning_results[i].mac_addr_str,
+				       latest_scan_results[i].mac_addr_str);
+				strcpy(request.scanning_results[i].ssid_str,
+				       latest_scan_results[i].ssid_str);
+				request.scanning_results[i].channel =
+					latest_scan_results[i].channel;
+				request.scanning_results[i].rssi =
+					latest_scan_results[i].rssi;
+			}
+
+			err = rest_services_wifi_location_get(wifi_config.service, &request,
+							      &result);
+			if (err) {
+				LOG_ERR("Failed to acquire a location by using "
+					"WiFi positioning, err: %d",
+						err);
+				err = -1;
+			} else {
+				location_result.latitude = result.latitude;
+				location_result.longitude = result.longitude;
+				location_result.accuracy = result.accuracy;
+				if (running) {
+					loc_core_event_cb(&location_result);
+					running = false;
+				}
+			}
+		} else {
+			if (latest_scan_result_count == 1) {
+				/* Following statement seems to be true at least with Here
+				 * (400: bad request) and also with Skyhook (404: not found).
+				 * Thus, fail faster in this case and save the data transfer costs.
+				 */
+				LOG_WRN("Retrieving a location based on a single WiFi access point "
+					"is not possible");
```

Comment:
@jayteemo @plskeggs Do you have the same understanding also that only one WiFi AP is not worth sending to the cloud?

Reply:
I'm just copying Pete's comment added elsewhere in this PR this same conversation:
---------------------
@trantanen I think WiFi is intended to be used for coarse positioning, similar to cell positioning, so one AP seems worth it. I'll check with Cole.

Answer: the current vendor has a lower limit of 2 APs, but other vendors could accept 1. I suggest making this a Kconfig value defaulting to 2.

Reply:
@plskeggs It seems currently all service providers that we have tried are returning an error for one AP and the REST response is saying something like "More than 1 AP needs to be given due to privacy reasons" so it seems there are privacy reasons why 1 AP cannot be accepted.
I won't be changing anything now unless we know a vendor that supports 1 AP. We can do the changes when that occurs if ever.

Reply:
OK